[unitaryHACK] `Register.hexagon()`, `Register.max_connectivity()` (#132), Enhanced `ValueError` messages (#148), Type hinting for `Register` (#16)

LaurentAjdnik opened the pull request:
Adds `Register.hexagon()` to generate a multi-layered hexagonal layout around a central atom.

Adds `Register.max_connectivity()` (#132) to generate a layout with maximal connectivity, based on a hexagonal layout as big as possible, and extra atoms placed in such a way that C3 and C6 rotational symmetries are enforced as often as possible.

Adds helper function `Register.__hexagon_full_layers()` used by both previous functions.

Adds enhanced `ValueError` messages (#148) for all functions in `Register`.

Adds TONS of tests about everything aforementioned.

Adds type hinting for all functions in `Register` (#16).

All tests succeed, with 100 % coverage.

Note: I might not be completely in the standards, considering I had NEVER EVER typed a single line of Python code before this PR. :grin::grin::grin:

LaurentAjdnik commented:
"build / style" failed because of "never used variables" in test functions? Really?... 🥺

What's your advice here ? Should I `# noqa: F841` the corresponding lines ?

HGSilveri commented:
Hi @LaurentAjdnik , first of all, thanks a lot for the PR! The style error you're getting is quite easy to fix, simply don't assign the `Register()` to `reg` when you're testing for errors. I'll make a comment on an example of this.

HGSilveri commented on pulser/tests/test_register.py:
```diff
@@ -62,6 +63,154 @@ def test_creation():
+def test_rectangle():
+    # Check rows
+    with pytest.raises(ValueError, match="The number of rows"):
+        reg = Register.rectangle(0, 2)
```
Since you'll never use `reg`, just do this instead:
```suggestion
       Register.rectangle(0, 2)
```

LaurentAjdnik commented:
> Hi @LaurentAjdnik , first of all, thanks a lot for the PR! The style error you're getting is quite easy to fix, simply don't assign the `Register()` to `reg` when you're testing for errors. I'll make a comment on an example of this.

So obvious! Silly me! Rookie mistake... Code updated.

HGSilveri reviewed (commented):
This looks great, very nice job! Here are some comments, all of them minor.

LaurentAjdnik commented:
Thanks so much for your feedback!

> As specified in PEP 484, the return type of functions or methods that don't return anything (like `__init__`) should be `None`, since the default in the absence of an annotation is `Any`.

MyPy implicitly treats the return type of `__init__` as `None` since [this PR](https://github.com/python/mypy/pull/5677). Hence, adding `-> None` would be irrelevant here ?


> The second check is redundant, because it is in rectangle too. This partially applies to the first check, if we don't care whether a distinction is made between "side" and "rows" or "columns", I leave that up to you to decide.

I guess it will depend on what we decide for #157.

I would favor checking everything first, before calling anything else, and return specific error messages, including the name of the field (`side` and not `rows`).

> Since you don't use cls, I think this would be better as a staticmethod

Done :+1:.

> I would rather have a single leading underscore on the name

Done :+1:.

> I just realized that the other docstrings lack a Returns field here too. Definitely not your job to add it everywhere, just please add it here.

Done for all functions in `Register` :+1:.

> Forgotten line from other docstring

Oh my! Removed :+1:.

> nit: Please remove the space after the @

Oh my again! Removed :+1:.

> Nonetheless, they way you're using the start_ and delta_ arrays, they could just be simple lists.

Indeed! Done :+1:.

> This is actually just atoms_left %= 6

Indeed! Done :+1:.

> It kinda sucks that you have to repeat all this code... Have you considered having this functionality built-in the __hexagon_full_layers helper (I guess the name would no longer apply), perhaps conditionally on being given values for atoms_left?

This one's gonna need a little bit of refactoring. Working on it :wink:.

> I guess you can type the `property` too, no?

Done :+1:. But not very precisely...

> Why the `return`?

Well... no idea :flushed:! Removed :+1:.

LaurentAjdnik commented:
> > Since you don't use cls, I think this would be better as a staticmethod
> 
> Done 👍.

Back to `@classMethod` after refactoring
 
> > It kinda sucks that you have to repeat all this code... Have you considered having this functionality built-in the __hexagon_full_layers helper (I guess the name would no longer apply), perhaps conditionally on being given values for atoms_left?
> 
> This one's gonna need a little bit of refactoring. Working on it 😉.

Done 👍.

HGSilveri commented on pulser/register.py:
```diff
@@ -96,6 +99,9 @@ def rectangle(cls, rows: int, columns: int, spacing: float = 4.0,
                 (e.g. prefix='q' -> IDs: 'q0', 'q1', 'q2', ...)
+
+        Returns:
+            Register: A register with qubits placed in a rectangular way.
```
nit: I "rectangular array" might be more appropriate. The same goes for the other docstrings.

HGSilveri commented on pulser/register.py:
```diff
@@ -107,10 +130,21 @@ def square(cls, side, spacing=4, prefix=None):
                 id starts with the prefix, followed by an int from 0 to N-1
                 (e.g. prefix='q' -> IDs: 'q0', 'q1', 'q2', ...).
         """
+        # Check side
+        if side < 1:
+            raise ValueError(
+                f"The number of atoms per side ({side}) must be 1 or above.")
+
+        # Check spacing
+        if spacing <= 0.0:
+            raise ValueError(f"Spacing ({spacing}) must be above 0.0.")
+
```
Just a reminder to address this

HGSilveri commented on pulser/register.py:
```diff
@@ -135,7 +185,169 @@ def triangular_lattice(cls, rows, atoms_per_row, spacing=4, prefix=None):
+        coords = cls.__hexagon_full_layers(layers, spacing)
+        return cls.from_coordinates(coords, center=False, prefix=prefix)
+
+    @ classmethod
```
Just a reminder to address this.

HGSilveri reviewed (commented):
This looks really good! Only some nits left to address

HGSilveri reviewed (approved):
Really great job, thanks a lot!
